ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,468 @@
+        storeConfig.zkClientConnectString, storeConfig.rootPath);
+    rootPath = storeConfig.rootPath;
+    List<String> subscribedPaths = Collections.singletonList(rootPath + FULL_ACCOUNT_METADATA_PATH);
+    helixStore = storeFactory.getHelixPropertyStore(storeConfig, subscribedPaths);
+    readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH);
+    logger.info("HelixAccountService started. {} accounts have been loaded, took time={}ms",
+        cache.getUniqueMapPair().getIdToAccountMap().values().size(), System.currentTimeMillis() - startTimeMs);
+  }
+
+  @Override
+  public Account getAccountByName(String accountName) {
+    if (accountName == null) {
+      throw new IllegalArgumentException("accountName cannot be null.");
+    }
+    return cache.getAccountByName(accountName);
+  }
+
+  @Override
+  public Account getAccountById(short id) {
+    return cache.getAccountById(id);
+  }
+
+  @Override
+  public Collection<Account> getAllAccounts() {
+    return cache.getAllAccounts();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   *   This call is blocking until it completes the operation of updating account metadata to {@link HelixPropertyStore}.
+   *   Returning {@code true} indicates that the accounts have been successfully written into {@link HelixPropertyStore},
+   *   {@code false} otherwise.
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
+    if (accounts == null) {
+      throw new IllegalArgumentException("accounts cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start updating to HelixPropertyStore with accounts={}", accounts);
+    checkDuplicateAccountIdOrName(accounts);
+    // make a pre check for conflict between the accounts to update and the accounts in cache. Will fail this update
+    // the update operation if any conflict exists. There is a slight chance that the account to update conflicts with
+    // the accounts in the local cache, but does not conflict with those in the helixPropertyStore. This will happen
+    // if some accounts are updated but the local cache is not refreshed.
+    Pair<Account, Account> conflictAccountPair = getConflictAccountWithCache(accounts);
+    if (conflictAccountPair != null) {
+      // @todo should we just log the case without throwing an exception? This case is indeed possible and may not be
+      // @todo caused by a programming error.
+      throw new IllegalArgumentException(
+          "Account to update conflicts with the existing account. accountId=" + conflictAccountPair.getFirst().getId()
+              + " accountName=" + conflictAccountPair.getFirst().getName() + " conflicts with existing accountId="
+              + conflictAccountPair.getSecond().getId() + " accountName=" + conflictAccountPair.getSecond().getName());
+    }
+
+    boolean res = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
```

Comment:
We have discussed offline and eventually we decided to keep it as is because the entire part is quite self-contained and we need to access some outside variables.